False positive ``missing-return-doc`` when a function is not supposed to be documented

Issue #2738 appears to persist on pylint 2.9.5 as well as 3.0.0-a4 (as retrieved using `pip install --upgrade [--pre] pylint` on this date). The merge request for that issue seems to imply this fix should have gone out with pylint 2.7.x.

Has something gone wrong here, or have I misunderstood the merge?

### Steps to reproduce

File to be analyzed:
```python
'''module docstring goes here'''

# this function doesn't require a docstring, because its name starts
# with an '_' (no-docstring-rgx):
def _function(some_arg: int) -> int:
    _ = some_arg
    return 0
```

rcfile:
```
[MASTER]
load-plugins=pylint.extensions.docparams

[PARAMETER DOCUMENTATION]
accept-no-param-doc=no
accept-no-return-doc=no
```

### Current behaviour

```
************* Module scrap
scrap.py:5:0: W9011: Missing return documentation (missing-return-doc)

------------------------------------------------------------------
Your code has been rated at 6.67/10 (previous run: 6.67/10, +0.00)
```
### Expected behavior
 
There should be no errors; `_function` requires no docstring and should therefore require no return documentation.

### pylint --version output

Result of `pylint --version` output:

```
pylint 3.0.0-a4
astroid 2.6.5
Python 3.8.3 (default, Jul  6 2020, 09:12:34)
[Clang 10.0.1 (clang-1001.0.46.4)]
```

--- a/doc/whatsnew/fragments/4743.bugfix
+++ b/doc/whatsnew/fragments/4743.bugfix
@@ -1,3 +1,4 @@
-``missing-return-doc`` now respects the ``no-docstring-rgx`` option.
+``missing-return-doc``, ``missing-raises-doc`` and ``missing-yields-doc`` now respect
+the ``no-docstring-rgx`` option.
 
 Closes #4743

--- a/pylint/extensions/docparams.py
+++ b/pylint/extensions/docparams.py
@@ -278,6 +278,11 @@ class DocstringParameterChecker(BaseChecker):
         if not isinstance(func_node, astroid.FunctionDef):
             return
 
+        # skip functions that match the 'no-docstring-rgx' config option
+        no_docstring_rgx = self.linter.config.no_docstring_rgx
+        if no_docstring_rgx and re.match(no_docstring_rgx, func_node.name):
+            return
+
         expected_excs = utils.possible_exc_types(node)
 
         if not expected_excs:
@@ -353,6 +358,11 @@ class DocstringParameterChecker(BaseChecker):
 
         func_node = node.frame(future=True)
         if not isinstance(func_node, astroid.FunctionDef):
+            return
+
+        # skip functions that match the 'no-docstring-rgx' config option
+        no_docstring_rgx = self.linter.config.no_docstring_rgx
+        if no_docstring_rgx and re.match(no_docstring_rgx, func_node.name):
             return
 
         doc = utils.docstringify(

--- a/tests/functional/ext/docparams/raise/missing_raises_doc_required.py
+++ b/tests/functional/ext/docparams/raise/missing_raises_doc_required.py
@@ -6,3 +6,10 @@
 def test_warns_unknown_style(self):  # [missing-raises-doc]
     """This is a docstring."""
     raise RuntimeError("hi")
+
+
+# This function doesn't require a docstring, because its name starts
+# with an '_' (no-docstring-rgx):
+def _function(some_arg: int):
+    """This is a docstring."""
+    raise ValueError

--- a/tests/functional/ext/docparams/yield/missing_yield_doc_required.py
+++ b/tests/functional/ext/docparams/yield/missing_yield_doc_required.py
@@ -1,7 +1,15 @@
 """Tests for missing-yield-doc and missing-yield-type-doc with accept-no-yields-doc = no"""
 # pylint: disable=missing-function-docstring, unused-argument, function-redefined
 
+from typing import Iterator
+
 
 # Test missing docstring
 def my_func(self):  # [missing-yield-doc, missing-yield-type-doc]
     yield False
+
+
+# This function doesn't require a docstring, because its name starts
+# with an '_' (no-docstring-rgx):
+def _function(some_arg: int) -> Iterator[int]:
+    yield some_arg

--- a/tests/functional/ext/docparams/yield/missing_yield_doc_required.txt
+++ b/tests/functional/ext/docparams/yield/missing_yield_doc_required.txt
@@ -1,2 +1,2 @@
-missing-yield-doc:6:0:6:11:my_func:Missing yield documentation:HIGH
-missing-yield-type-doc:6:0:6:11:my_func:Missing yield type documentation:HIGH
+missing-yield-doc:8:0:8:11:my_func:Missing yield documentation:HIGH
+missing-yield-type-doc:8:0:8:11:my_func:Missing yield type documentation:HIGH

